File: apps/mobile/modules/share-save/android/src/main/java/app/allkept/sharesave/ShareSaveModule.kt
package app.allkept.sharesave

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ShareSaveModule : Module() {
  private val context get() = requireNotNull(appContext.reactContext) { "React context is not ready" }

  override fun definition() = ModuleDefinition {
    Name("ShareSave")
    Function("setCredential") { json: String -> SharedStore.setCredential(context, json); ShareShortcut.publish(context) }
    Function("clearCredential") { SharedStore.clearCredential(context); ShareShortcut.remove(context) }
    Function("hasCredential") { SharedStore.credential(context) != null }
    Function("peekQueue") { SharedStore.queueJson(context) }
    Function("dropQueued") { requestId: String -> SharedStore.drop(context, requestId) }
  }
}

File: apps/mobile/modules/share-save/ios/ShareSaveModule.swift
import ExpoModulesCore

public class ShareSaveModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ShareSave")

    Function("setCredential") { (json: String) throws in
      let credential = try JSONDecoder().decode(SharedStore.Credential.self, from: Data(json.utf8))
      try SharedStore.setCredential(credential)
    }
    Function("clearCredential") { SharedStore.clearCredential() }
    Function("hasCredential") { () -> Bool in SharedStore.credential() != nil }
    Function("peekQueue") { () -> String in
      let data = (try? JSONEncoder().encode(SharedStore.queue())) ?? Data("[]".utf8)
      return String(decoding: data, as: UTF8.self)
    }
    Function("dropQueued") { (requestId: String) in SharedStore.drop(requestId: requestId) }
  }
}
